### PR TITLE
4.10 to 4.11 upgrade path

### PR DESCRIPTION
**Pull request by DaanHoogland**

### CODE REVIEW

File: engine/schema/src/com/cloud/upgrade/DatabaseUpgradeChecker.java [hunk 1]
```diff
@@ -104,210 +105,353 @@
...
-    protected Map<CloudStackVersion, DbUpgrade[]> _upgradeMap = new HashMap<>();
-
-    @Inject
-    VersionDao _dao;
-
     private final ImmutableList<CloudStackVersion> availableVersions;
+    protected Map<CloudStackVersion, DbUpgrade[]> _upgradeMap = new HashMap<>();
+    @Inject VersionDao _dao;
 
     public DatabaseUpgradeChecker() {
```
Comment (rhtyd) [inline engine/schema/src/com/cloud/upgrade/DatabaseUpgradeChecker.java]: @DaanHoogland it will be hard to review the changes as there as all the newlines and formating has changed. Kindly revert some of that, for example having the `@Annotations` above the variable declarations.

Comment (DaanHoogland) [inline engine/schema/src/com/cloud/upgrade/DatabaseUpgradeChecker.java]: split the reformat into a separate commit

Comment (DaanHoogland) [inline engine/schema/src/com/cloud/upgrade/DatabaseUpgradeChecker.java]: ping @rhtyd

File: engine/schema/src/com/cloud/upgrade/dao/Upgrade41000to41100.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+import com.cloud.utils.script.Script;
+import org.apache.log4j.Logger;
+
+import java.io.File;
+import java.sql.Connection;
+
+public class Upgrade41000to41100 implements DbUpgrade {
+    final static Logger LOG = Logger.getLogger(Upgrade41000to41100.class);
+
+    public static class MemoryValues {
```
Comment (rhtyd) [inline engine/schema/src/com/cloud/upgrade/dao/Upgrade41000to41100.java]: @DaanHoogland why do we have this class here?

**Comment by rhtyd** [comment]
Sorry !LGTM, @DaanHoogland please remove unnecessary classes from 410->411 upgrade path class. Remove unnecessary formatting, makes harder to review backward compatibility issues. We need to test for a fresh installation which path runs first 4.1->4.2 or 4.10/4.11 since I suspect there could be a slight db upgrade issue when 4.1 and 4.10 version strings are compared (it might be possible that we don't have that issue anymore).

**Comment by cloudmonger** [comment]
### ACS CI BVT Run
 **Sumarry:**
 Build Number 958
 Hypervisor xenserver
 NetworkType Advanced
 Passed=105
 Failed=8
 Skipped=12

_Link to logs Folder (search by build_no):_ https://www.dropbox.com/sh/r2si930m8xxzavs/AAAzNrnoF1fC3auFrvsKo_8-a?dl=0

**Failed tests:**
* test_non_contigiousvlan.py

 * test_extendPhysicalNetworkVlan Failed

* test_deploy_vm_iso.py

 * test_deploy_vm_from_iso Failing since 19 runs

* test_volumes.py

 * test_06_download_detached_volume Failing since 3 runs

* test_vm_life_cycle.py

 * test_10_attachAndDetach_iso Failing since 19 runs

* test_routers_network_ops.py

 * test_01_isolate_network_FW_PF_default_routes_egress_true Failing since 52 runs

 * test_02_isolate_network_FW_PF_default_routes_egress_false Failing since 52 runs

 * test_01_RVR_Network_FW_PF_SSH_default_routes_egress_true Failing since 52 runs

 * test_02_RVR_Network_FW_PF_SSH_default_routes_egress_false Failing since 52 runs


**Skipped tests:**
test_vm_nic_adapter_vmxnet3
test_01_verify_libvirt
test_02_verify_libvirt_after_restart
test_03_verify_libvirt_attach_disk
test_04_verify_guest_lspci
test_05_change_vm_ostype_restart
test_06_verify_guest_lspci_again
test_static_role_account_acls
test_11_ss_nfs_version_on_ssvm
test_nested_virtualization_vmware
test_3d_gpu_support
test_deploy_vgpu_enabled_vm

**Passed test suits:**
test_deploy_vm_with_userdata.py
test_affinity_groups_projects.py
test_portable_publicip.py
test_vm_snapshots.py
test_over_provisioning.py
test_global_settings.py
test_scale_vm.py
test_service_offerings.py
test_routers_iptables_default_policy.py
test_loadbalance.py
test_routers.py
test_reset_vm_on_reboot.py
test_deploy_vms_with_varied_deploymentplanners.py
test_network.py
test_router_dns.py
test_login.py
test_list_ids_parameter.py
test_public_ip_range.py
test_multipleips_per_nic.py
test_metrics_api.py
test_regions.py
test_affinity_groups.py
test_network_acl.py
test_pvlan.py
test_nic.py
test_deploy_vm_root_resize.py
test_resource_detail.py
test_secondary_storage.py
test_disk_offerings.py

**Comment by DaanHoogland** [comment]
@rhtyd of course I will remove that unnecessary class. it is a c&p bug. I will *not* remove the formatting. I put it in a separate bug to ease reviewing. It is in no way in your way and it improves the readability of the code.

**Comment by rhtyd** [comment]
LGTM, tested locally. Thanks @DaanHoogland

**Comment by rhtyd** [comment]
mysql> use cloud;mysql> select * from version;
+----+----------+---------------------+----------+
| id | version  | updated             | step     |
+----+----------+---------------------+----------+
|  1 | 4.0.0    | 2017-07-13 15:41:33 | Complete |
|  2 | 4.1.0    | 2017-07-13 10:11:53 | Complete |
|  3 | 4.2.0    | 2017-07-13 10:12:06 | Complete |
|  4 | 4.2.1    | 2017-07-13 10:12:06 | Complete |
|  5 | 4.3.0    | 2017-07-13 10:12:12 | Complete |
|  6 | 4.4.0    | 2017-07-13 10:12:20 | Complete |
|  7 | 4.4.1    | 2017-07-13 10:12:20 | Complete |
|  8 | 4.4.2    | 2017-07-13 10:12:20 | Complete |
|  9 | 4.5.0    | 2017-07-13 10:12:24 | Complete |
| 10 | 4.5.1    | 2017-07-13 10:12:24 | Complete |
| 11 | 4.5.2    | 2017-07-13 10:12:25 | Complete |
| 12 | 4.6.0    | 2017-07-13 10:12:26 | Complete |
| 13 | 4.6.1    | 2017-07-13 10:12:26 | Complete |
| 14 | 4.7.0    | 2017-07-13 10:12:27 | Complete |
| 15 | 4.7.1    | 2017-07-13 10:12:27 | Complete |
| 16 | 4.8.0    | 2017-07-13 10:12:27 | Complete |
| 17 | 4.8.1    | 2017-07-13 10:12:27 | Complete |
| 18 | 4.9.0    | 2017-07-13 10:12:30 | Complete |
| 19 | 4.9.1.0  | 2017-07-13 10:12:31 | Complete |
| 20 | 4.9.2.0  | 2017-07-13 10:12:31 | Complete |
| 21 | 4.9.3.0  | 2017-07-13 10:12:31 | Complete |
| 22 | 4.10.0.0 | 2017-07-13 10:12:32 | Complete |
| 23 | 4.11.0.0 | 2017-07-13 10:12:33 | Complete |
+----+----------+---------------------+----------+
